safemysql.class.php

```diff
@@ -517,35 +554,66 @@ public function getStats()
 	/**
 	 * private function which actually executes $query.
 	 * also logs some stats like profiling info and error message
-	 * 
+	 *
 	 * @param string $query - a regular SQL query
 	 * @return mysqli result resource or FALSE on error
 	 */
 	private function rawQuery($query)
 	{
-		$start = microtime(TRUE);
-		$res   = mysqli_query($this->conn, $query);
-		$timer = microtime(TRUE) - $start;
+		return $this->retryIfDeadlocked(
+			function($db) use($query)
+			{
+				$start = microtime(TRUE);
+				$res   = mysqli_query($this->conn, $query);
+				$timer = microtime(TRUE) - $start;
+
+				$this->stats[] = array(
+					'query' => $query,
+					'start' => $start,
+					'timer' => $timer,
+				);
+				if (!$res)
+				{
+					$error = mysqli_error($this->conn);
+					$errno = mysqli_errno($this->conn);
 
-		$this->stats[] = array(
-			'query' => $query,
-			'start' => $start,
-			'timer' => $timer,
+					end($this->stats);
+					$key = key($this->stats);
+					$this->stats[$key]['error'] = $error;
+					$this->cutStats();
+
+					$this->error("$error. Full query: [$query]", $errno);
+				}
+				$this->cutStats();
+				return $res;
+			}
 		);
-		if (!$res)
+	}
+
+	private function retryIfDeadlocked($callback, $can_retry = null)
```

Comment:
- [ ] The naming here is horrible; the function claims to retry if deadlocked, but then you're optionally passing a parameter whose whole purpose is to make it *not* do that.